test/Concurrency/Runtime/async_sequence.swift

```diff
@@ -0,0 +1,524 @@
+}
+
+var AsyncDropWhileSequenceTests = TestSuite("AsyncDropWhileSequence")
+
+AsyncDropWhileSequenceTests.test("drop one") {
+  let results = await [1, 2, 3].async.drop { value -> Bool in
+    return value == 1
+  }.collect()
+  expectEqual(results, [2, 3])
+}
+
+AsyncDropWhileSequenceTests.test("drop one async") {
+  let results = await [1, 2, 3].async.drop { value async -> Bool in
+    return await asynchronous(value == 1)
+  }.collect()
+  expectEqual(results, [2, 3])
+}
+
+AsyncDropWhileSequenceTests.test("drop one throws") {
+  do {
+    let results = try await [1, 2, 3].async.drop { value throws -> Bool in
+      return try throwing(value == 1)
+    }.collect()
+    expectEqual(results, [2, 3])
+  } catch {
+    expectUnreachable()
+  }
+}
+
+AsyncDropWhileSequenceTests.test("drop one throws async") {
+  do {
+    let results = try await [1, 2, 3].async.drop { value async throws -> Bool in
+      return try await asynchronousThrowing(value == 1)
+    }.collect()
+    expectEqual(results, [2, 3])
+  } catch {
+    expectUnreachable()
+  }
+}
+
+AsyncDropWhileSequenceTests.test("drop none") {
+  let results = await [1, 2, 3].async.drop { _ in return false }.collect()
+  expectEqual(results, [1, 2, 3])
+}
+
+AsyncDropWhileSequenceTests.test("drop all") {
+  let results = await [1, 2, 3].async.drop { _ in return true }.collect()
+  expectEqual(results, [])
+}
+
+AsyncDropWhileSequenceTests.test("drop stops") {
+  let results = await [1, 2, 3].async.drop { value in
+    return value == 1 || value == 3
+  }.collect()
+  expectEqual(results, [2, 3])
+}
+
+var AsyncDropFirstSequenceTests = TestSuite("AsyncDropFirstSequence")
+
+AsyncDropFirstSequenceTests.test("drop 1") {
+  let results = await [1, 2, 3].async.dropFirst().collect()
+  expectEqual(results, [2, 3])
+}
+
+AsyncDropFirstSequenceTests.test("drop 2") {
+  let results = await [1, 2, 3].async.dropFirst(2).collect()
+  expectEqual(results, [3])
+}
+
+AsyncDropFirstSequenceTests.test("drop 0") {
+  let results = await [1, 2, 3].async.dropFirst(0).collect()
+  expectEqual(results, [1, 2, 3])
+}
+
+var AsyncPrefixSequenceTests = TestSuite("AsyncPrefixSequence")
+
+AsyncPrefixSequenceTests.test("prefix satisfied") {
+  let results = await [1, 2, 3].async.prefix(2).collect()
+  expectEqual(results, [1, 2])
+}
+
+AsyncPrefixSequenceTests.test("prefix partial satisfied") {
+  let results = await [1, 2, 3].async.prefix(4).collect()
+  expectEqual(results, [1, 2, 3])
+}
+
+AsyncPrefixSequenceTests.test("prefix 0") {
+  let results = await [1, 2, 3].async.prefix(0).collect()
+  expectEqual(results, [])
+}
+
+var AsyncPrefixWhileSequenceTests = TestSuite("AsyncPrefixWhileSequence")
+
+/* DISABLED: this deadlocks for some unknown reason
+AsyncPrefixWhileSequenceTests.test("prefix while") {
+  let results = await [1, 2, 3].async
+    .prefix { $0 < 3 }
+    .collect()
+  expectEqual(results, [1, 2])
+}
+*/
+
+var AsyncFlatMapSequenceTests = TestSuite("AsyncFlatMapSequence")
```

Comment:
Need tests for throwing closures that throw and don't throw.

Reply:
yea, and we need to add tests on upstream failures.

BTW... testing the throwing paths for AsyncFlatMapSequence actually exposed some bugs :X